dagstore_gc.go

```diff
@@ -0,0 +1,101 @@
+package dagstore
+
+import (
+	"io/fs"
+	"os"
+	"path/filepath"
+
+	"github.com/filecoin-project/dagstore/shard"
+)
+
+// GCResult is the result of performing a GC operation. It holds the results
+// from deleting unused transients, and the result from reconciling the
+// transients dir with files referenced from mounts.
+type GCResult struct {
+	// Shards includes an entry for every shard whose transient was reclaimed.
+	// Nil error values indicate success.
+	Shards map[shard.Key]error
+	// Reconcile stores the result from the reconciliation phase.
+	Reconcile error
+}
+
+// ShardFailures returns the number of shards whose transient reclaim failed.
+func (e *GCResult) ShardFailures() int {
+	var failures int
+	for _, err := range e.Shards {
+		if err != nil {
+			failures++
+		}
+	}
+	return failures
+}
+
+// gc performs DAGStore GC. Refer to DAGStore#GC for more information.
+//
+// The event loops gives it exclusive execution rights, so while GC is running,
+// no other events are being processed.
+func (d *DAGStore) gc(resCh chan *GCResult) {
+	res := &GCResult{
+		Shards: make(map[shard.Key]error),
+	}
+
+	// A. Delete transients for unused and errored shards.
+
+	// determine which shards can be reclaimed.
+	d.lk.RLock()
+	var reclaim []*Shard
+	for _, s := range d.shards {
+		s.lk.RLock()
+		if nAcq := len(s.wAcquire); (s.state == ShardStateAvailable || s.state == ShardStateErrored) && nAcq == 0 {
+			reclaim = append(reclaim, s)
+		}
+		s.lk.RUnlock()
+	}
+	d.lk.RUnlock()
+
+	// attempt to delete transients of reclaimed shards.
+	for _, s := range reclaim {
+		s.lk.Lock()
+		err := s.mount.DeleteTransient()
+		if err != nil {
+			log.Warnw("failed to delete transient", "shard", s.key, "error", err)
+		}
+
+		// record the error so we can return it.
+		res.Shards[s.key] = err
+
+		// flush the shard state to the datastore.
+		if err := s.persist(d.config.Datastore); err != nil {
+			log.Warnw("failed to persist shard", "shard", s.key, "error", err)
+		}
+		s.lk.Unlock()
+	}
+
+	// B. Remove files that aren't referenced by the mounts.
```

Comment:
I don't think we need to do this reconciliation on every GC. Just doing it on startup should be fine. We'll have unreferenced files only if we crash before failing to persist the transient path or crash before removing a partial file on a fetch error or crash before we mark a partial as a transient. 

This is adding needless cognitive overhead as I just had to reason about how this could race with the Upgrader to convince myself that it wont.

Reply:
That makes sense. It would allow us to get rid of exposing the partial file path too, and the associated complexity. We had discussed this, but I completely forgot.

Reply:
On second thought, I don’t think I want the reconciliation to run only on startup. It’s really valuable to have it run in normal GC, as it makes the system more resilient, e.g. imagine we’re stuck in a loop where fetching a 32GiB mount fails halfway, repeatedly. We’d keep accumulating garbage that would only get collected on startup. Also, in the future, we want to set a disk quota for the transients dir, and have GC kick off based on a watermark policy.

Reply:
I've pushed a commit to simplify the solution. Since now we're using precomputed filenames for complete and partial files, and not generating random ones, the risk of us accumulating garbage due to thrash is very low/nil. So now we can rely on orphan cleanup on startup only.